### ISSUE TITLE
FastR unexpected failure: R_HOME is not set correctly

### ISSUE THREAD

**Issue by thiagohora** [oracle/graal#1322]
Hey, I'm getting this error when trying to init the context of R:

```
Caused by: org.graalvm.polyglot.PolyglotException
	at org.graalvm.polyglot.Context.eval(Context.java:367)
	at com.glovoapp.hackathon.r.FastR.eval(FastR.java:171)
	at com.glovoapp.hackathon.r.FastR.<init>(FastR.java:56)
	at com.glovoapp.hackathon.r.FastR.<init>(FastR.java:51)
	at com.glovoapp.hackathon.r.FastRConfig.fastR(FastRConfig.java:11)
```

```
 public FastR() {
        this(Context.newBuilder("R").allowAllAccess(true).build());
    }
```
```
[dev-thiago@localhost R]$ ./bin/R
R version 3.5.1 (FastR)
Copyright (c) 2013-19, Oracle and/or its affiliates
Copyright (c) 1995-2018, The R Core Team
Copyright (c) 2018 The R Foundation for Statistical Computing
Copyright (c) 2012-4 Purdue University
Copyright (c) 1997-2002, Makoto Matsumoto and Takuji Nishimura
All rights reserved.

FastR is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information.

Type 'q()' to quit R.
FastR unexpected failure: R_HOME is not set correctly
```

**Comment by steve-s** [comment]
Hello Thiago, 

are you building a native-image of the Java application where you are embedding FastR? FastR needs to know where on the filesystem it should look for base libraries, configuration, and other necessary files. If you run FastR in JVM mode, then it can infer this from the location of `fastr.jar`. In the native-image mode, you need to export environment variable `R_HOME` pointing to the location of FastR home in `$GRAALVM_HOME/jre/languages/R`.

TL;DR:
```
R_HOME=$GRAALVM_HOME/jre/languages/R ./bin/R
```
should fix the issue. Please re-open if it doesn't.

The error message has already been improved in the development version: https://github.com/oracle/fastr/commit/d982169145fea7eb7e52414ca98ad442386b0b51.

**Comment by thiagohora** [comment]
I'm running in my IDE in JVM mode and I'm getting the same error. The fastr.jar must be in the classpath or there are another configuration needed?

**Comment by steve-s** [comment]
On top of creating a project from "hello world" template, I only needed to add `jre/lib/boot/graal-sdk.jar` to the class-path so that IntelliJ can see the Graal SDK classes. Following example compiled and run:
```
import org.graalvm.polyglot.Context;

public class Main {

    public static void main(String[] args) {
        Context ctx = Context.newBuilder("R").allowAllAccess(true).build();
        ctx.eval("R", "print(c(2,3) + c(3+5))");
    }
}
```
Here is my class-path in IntelliJ:
![idea-classpath](https://user-images.githubusercontent.com/1492237/58825675-41696a80-863f-11e9-91d6-c87078e2aff8.png)

Note that you don't really need to add `jre/lib/boot/graal-sdk.jar` on the class-path to run the application. It is added for you by GraalVM, but IntelliJ doesn't seem to recognize that. You can check this example of FastR embedding: https://github.com/graalvm/examples/tree/master/r_java_embedding

**Comment by thiagohora** [comment]
Yes, I took this project as example and in the mac of my friend it was working. But on my linux it is not.

Trying to run this simple main example I got this:

```
FastR unexpected failure: R_HOME is not set correctly
Exception in thread "main" org.graalvm.polyglot.PolyglotException
	at org.graalvm.polyglot.Context.eval(Context.java:367)
	at com.glovoapp.hackathon.r.FastR.main(FastR.java:22)
```

**Comment by steve-s** [comment]
There is a known issue with `R_HOME` if GraalVM is installed in a directory with whitespace or any character that needs to be URL encoded, but judging from your screenshot, this is not your case.

Can you try running the example from https://github.com/graalvm/examples/tree/master/r_java_embedding outside of the IDE? You just need to `export GRAALVM_DIR=/usr/lib/jvm/graalvm-ce-19.0.0` and then run `./build.sh && ./run.sh`. Does that work?

**Comment by steve-s** [comment]
A workaround could be setting the `R_HOME` environment variable explicitly to `/usr/lib/jvm/graalvm-ce-19.0.0/jre/languages/R`.

**Comment by thiagohora** [comment]
I tried to set the the `R_HOME` in the environment and in the running option. 

![image](https://user-images.githubusercontent.com/6633270/58831437-02daac80-864d-11e9-9ee0-15f140eee67a.png)

But I will try to build the example (https://github.com/graalvm/examples/tree/master/r_java_embedding)

**Comment by thiagohora** [comment]
Any thoughts on that?

**Comment by steve-s** [comment]
Can you please check that `/usr/lib/jvm/graalvm-ce-19.0.0/jre/languages/R` contains a folder named `etc` and in that folder file `Makeconf` and both are readable for the user that executes the example? if that's not the case, try reinstalling R with the `gu` utility.

**Comment by thiagohora** [comment]
```
# Makeconf.  Generated from Makeconf.in by configure.
#
# ${R_HOME}/etc/Makeconf
#
# R was configured using the following call
# (not including env. vars and site configuration)
# configure  '--with-x=no' '--with-aqua=no' '--enable-memory-profiling' 'FFLAGS=-O2'

## This fails if it contains spaces, or if it is quoted
include $(R_SHARE_DIR)/make/vars.mk

AR = ar
BLAS_LIBS = -L"$(R_HOME)/lib$(R_ARCH)" -lRblas
C_VISIBILITY = -fvisibility=hidden
CC = gcc
CFLAGS = -g -O2 $(LTO)
CPICFLAGS = -fpic
CPPFLAGS = -I/usr/local/include
CXX = g++ 
CXXCPP = $(CXX) -E
CXXFLAGS = -g -O2 $(LTO)
CXXPICFLAGS = -fpic
CXX98 = g++
CXX98FLAGS = -g -O2
CXX98PICFLAGS = -fpic
CXX98STD = -std=gnu++98
CXX11 = g++
CXX11FLAGS = -g -O2
CXX11PICFLAGS = -fpic
CXX11STD = -std=gnu++11
CXX14 = g++
CXX14FLAGS = -g -O2
CXX14PICFLAGS = -fpic
CXX14STD = -std=gnu++14
CXX17 = g++
CXX17FLAGS = -g -O2
CXX17PICFLAGS = -fpic
CXX17STD = -std=gnu++17
DYLIB_EXT = .so
DYLIB_LD = $(CC)
DYLIB_LDFLAGS = -shared -fopenmp# $(CFLAGS) $(CPICFLAGS)
DYLIB_LINK = $(DYLIB_LD) $(DYLIB_LDFLAGS) $(LDFLAGS)
ECHO = echo
ECHO_C = 
ECHO_N = -n
ECHO_T = 
## NB, set FC before F77 as on Solaris make, setting FC sets F77
FC = gfortran
FCFLAGS = -g -O2 $(LTO)
## additional libs needed when linking with $(FC), e.g. on some Oracle compilers
FCLIBS = 
F77 = f95
F77_VISIBILITY = -fvisibility=hidden
FFLAGS = -O2 $(LTO)
FLIBS =  -lgfortran -lm -lquadmath
FCPICFLAGS = -fpic
FPICFLAGS = -fpic
FOUNDATION_CPPFLAGS = 
FOUNDATION_LIBS = 
JAR = 
JAVA = 
JAVAC = 
JAVAH = 
## JAVA_HOME might be used in the next three.  
## They are for packages 'JavaGD' and 'rJava'
JAVA_HOME = 
JAVA_CPPFLAGS = 
JAVA_LIBS = 
JAVA_LD_LIBRARY_PATH = 
LAPACK_LIBS = -L"$(R_HOME)/lib$(R_ARCH)" -lRlapack
## we only need this is if it is external, as otherwise link to R
LIBINTL= @LIBINTL@
LIBM = -lm
LIBR0 = 
LIBR1 = 
LIBR =  
LIBS =  -lpcre2-8 -lpcre -llzma -lbz2 -lz -lrt -ldl -lm -licuuc -licui18n
## needed by R CMD config
LIBnn = lib64
LIBTOOL = $(SHELL) "$(R_HOME)/bin/libtool"
LDFLAGS = -L/usr/local/lib64
LTO = 
## needed to build applications linking to static libR
MAIN_LD = $(CC)
MAIN_LDFLAGS = -Wl,--export-dynamic -fopenmp
MAIN_LINK = $(MAIN_LD) $(MAIN_LDFLAGS) $(LDFLAGS)
MKINSTALLDIRS = "$(R_HOME)/bin/mkinstalldirs"
OBJC = 
OBJCFLAGS =  $(LTO)
OBJC_LIBS = 
OBJCXX = 
R_ARCH = 
RANLIB = ranlib
SAFE_FFLAGS = -O2 -msse2 -mfpmath=sse
SED = /usr/bin/sed
SHELL = /bin/sh
SHLIB_CFLAGS = 
SHLIB_CXXFLAGS = 
SHLIB_CXXLD = $(CXX)
SHLIB_CXXLDFLAGS = -shared
SHLIB_CXX98LD = $(CXX98) $(CXX98STD)
SHLIB_CXX98LDFLAGS = -shared
SHLIB_CXX11LD = $(CXX11) $(CXX11STD)
SHLIB_CXX11LDFLAGS = -shared
SHLIB_CXX14LD = $(CXX14) $(CXX14STD)
SHLIB_CXX14LDFLAGS = -shared
SHLIB_CXX17LD = $(CXX17) $(CXX17STD)
SHLIB_CXX17LDFLAGS = -shared
SHLIB_EXT = .so
SHLIB_FCLD = $(FC)
SHLIB_FCLDFLAGS = -shared
SHLIB_FFLAGS = 
SHLIB_LD = $(CC)
SHLIB_LDFLAGS = -shared# $(CFLAGS) $(CPICFLAGS)
SHLIB_LIBADD = 
## We want to ensure libR is picked up from $(R_HOME)/lib
## before e.g. /usr/local/lib if a version is already installed.
SHLIB_LINK = $(SHLIB_LD) $(SHLIB_LDFLAGS) $(LIBR0) $(LDFLAGS)
SHLIB_OPENMP_CFLAGS = -fopenmp
SHLIB_OPENMP_CXXFLAGS = -fopenmp
SHLIB_OPENMP_FCFLAGS = -fopenmp
SHLIB_OPENMP_FFLAGS = -fopenmp
STRIP_LIBS = strip --strip-unneeded
STRIP_STATIC_LIBS = strip --strip-debug
TCLTK_CPPFLAGS = -I/usr/include -I/usr/include 
TCLTK_LIBS = -L/usr/lib64 -ltcl8.6 -L/usr/lib64 -ltk8.6 -lX11
YACC = yacc

## legacy
CXX1X = g++
CXX1XFLAGS = -g -O2
CXX1XPICFLAGS = -fpic
CXX1XSTD = -std=gnu++11
SHLIB_CXX1XLD = $(CXX11) $(CXX11STD)
SHLIB_CXX1XLDFLAGS = -shared


## for linking to libR.a
STATIC_LIBR = # -Wl,--whole-archive "$(R_HOME)/lib$(R_ARCH)/libR.a" -Wl,--no-whole-archive $(BLAS_LIBS) $(FLIBS)  $(LIBINTL) -lreadline  $(LIBS)

R_XTRA_CFLAGS = 
R_XTRA_CPPFLAGS =  -I"$(R_INCLUDE_DIR)" -DNDEBUG
R_XTRA_CXXFLAGS = 
R_XTRA_FFLAGS = 

ALL_CFLAGS = $(R_XTRA_CFLAGS) $(PKG_CFLAGS) $(CPICFLAGS) $(SHLIB_CFLAGS) $(CFLAGS)
ALL_CPPFLAGS = $(R_XTRA_CPPFLAGS) $(PKG_CPPFLAGS) $(CLINK_CPPFLAGS) $(CPPFLAGS)
ALL_CXXFLAGS = $(R_XTRA_CXXFLAGS) $(PKG_CXXFLAGS) $(CXXPICFLAGS) $(SHLIB_CXXFLAGS) $(CXXFLAGS)
ALL_OBJCFLAGS = $(PKG_OBJCFLAGS) $(CPICFLAGS) $(SHLIB_CFLAGS) $(OBJCFLAGS)
ALL_OBJCXXFLAGS = $(PKG_OBJCXXFLAGS) $(CXXPICFLAGS) $(SHLIB_CXXFLAGS) $(OBJCXXFLAGS)
ALL_FFLAGS = $(R_XTRA_FFLAGS) $(PKG_FFLAGS) $(FPICFLAGS) $(SHLIB_FFLAGS) $(FFLAGS)
## LIBR here as a couple of packages use this without SHLIB_LINK
ALL_LIBS = $(PKG_LIBS) $(SHLIB_LIBADD) $(LIBR)# $(LIBINTL)

.SUFFIXES:
.SUFFIXES: .c .cc .cpp .d .f .f90 .f95 .m .mm .M .o

.c.o:
	$(CC) $(ALL_CPPFLAGS) $(ALL_CFLAGS) -c $< -o $@
.c.d:
	@echo "making $@ from $<"
	@$(CC) -MM $(ALL_CPPFLAGS) $< > $@
.m.d:
	@echo > $@
.cc.o:
	$(CXX) $(ALL_CPPFLAGS) $(ALL_CXXFLAGS) -c $< -o $@
.cpp.o:
	$(CXX) $(ALL_CPPFLAGS) $(ALL_CXXFLAGS) -c $< -o $@
.cc.d:
	@echo "making $@ from $<"
	@$(CXX) -M $(ALL_CPPFLAGS) $< > $@
.cpp.d:
	@echo "making $@ from $<"
	@$(CXX) -M $(ALL_CPPFLAGS) $< > $@
.m.o:
	$(OBJC) $(ALL_CPPFLAGS) $(ALL_OBJCFLAGS) -c $< -o $@
.mm.o:
	$(OBJCXX) $(ALL_CPPFLAGS) $(ALL_OBJCXXFLAGS) -c $< -o $@
.M.o:
	$(OBJCXX) $(ALL_CPPFLAGS) $(ALL_OBJCXXFLAGS) -c $< -o $@
.f.o:
	$(F77) $(ALL_FFLAGS) -c $< -o $@
.f95.o:
	$(FC) $(PKG_FCFLAGS) $(FCPICFLAGS) $(FCFLAGS) -c  $< -o $@
.f90.o:
	$(FC) $(PKG_FCFLAGS) $(FCPICFLAGS) $(FCFLAGS) -c  $< -o $@

```

**Comment by thurka** [comment]
> configure_fastr: line 43: ed: command not found
It looks like you are missing `ed` in your linux installation. Can you please install `ed` and reinstall R.

**Comment by thiagohora** [comment]
HI @thurka you are right, I have installed ed and now it is working, thanks!

Also thank you @steve-s

**Comment by spannozzo** [comment]
I had the same issue when I was running the native image with sudo. the R_HOME variable must be set also into the root account
